vaadin-spring/src/main/java/com/vaadin/flow/spring/scopes/VaadinRouteScope.java

```diff
@@ -0,0 +1,380 @@
+import com.vaadin.flow.router.AfterNavigationEvent;
+import com.vaadin.flow.router.AfterNavigationListener;
+import com.vaadin.flow.router.BeforeEnterEvent;
+import com.vaadin.flow.router.BeforeEnterListener;
+import com.vaadin.flow.router.RouterLayout;
+import com.vaadin.flow.server.UIInitEvent;
+import com.vaadin.flow.server.UIInitListener;
+import com.vaadin.flow.server.VaadinContext;
+import com.vaadin.flow.server.VaadinService;
+import com.vaadin.flow.server.VaadinServletContext;
+import com.vaadin.flow.server.VaadinSession;
+import com.vaadin.flow.shared.Registration;
+import com.vaadin.flow.spring.SpringVaadinSession;
+import com.vaadin.flow.spring.annotation.RouteScopeOwner;
+
+/**
+ * Implementation of Spring's
+ * {@link org.springframework.beans.factory.config.Scope} that binds the beans
+ * to a component in the navigation chain. Registered by default as the scope "
+ * {@value #VAADIN_ROUTE_SCOPE_NAME}".
+ *
+ * @see com.vaadin.flow.spring.annotation.VaadinSessionScope
+ * @author Vaadin Ltd
+ * @since
+ *
+ */
+public class VaadinRouteScope extends AbstractScope implements UIInitListener {
+
+    public static final String VAADIN_ROUTE_SCOPE_NAME = "vaadin-route";
+
+    private static class RouteStoreWrapper implements Serializable {
+
+        private final VaadinSession session;
+
+        private final Registration sessionDestroyListenerRegistration;
+
+        private final Map<String, BeanStore> routeStores;
+
+        private RouteStoreWrapper(VaadinSession session) {
+            assert session.hasLock();
+            this.session = session;
+            routeStores = new HashMap<>();
+            if (session instanceof SpringVaadinSession) {
+                sessionDestroyListenerRegistration = null;
+                ((SpringVaadinSession) session)
+                        .addDestroyListener(event -> destroy());
+            } else {
+                sessionDestroyListenerRegistration = session.getService()
+                        .addSessionDestroyListener(event -> destroy());
+            }
+        }
+
+        private BeanStore getBeanStore(UI ui) {
+            assert session.hasLock();
+            ExtendedClientDetails details = ui.getInternals()
+                    .getExtendedClientDetails();
+            String key = getUIStoreKey(ui);
+            if (details == null) {
+                ui.getPage().retrieveExtendedClientDetails(
+                        det -> relocateStore(ui, key));
+            }
+            BeanStore beanStore = routeStores.get(key);
+            if (beanStore == null) {
+                beanStore = new RouteBeanStore(ui, session);
+                routeStores.put(key, beanStore);
+            }
+            return beanStore;
+        }
+
+        private void relocateStore(UI ui, String key) {
+            assert session.hasLock();
+            BeanStore beanStore = routeStores.remove(key);
+            if (beanStore == null) {
+                // TODO
```

Comment:
??? By minimum would need commenting on _what_ is to be done

Reply:
Leftover. Thanks.